PresentData/PresentMonTraceConsumer.cpp

```diff
@@ -1139,12 +1181,152 @@ void PMTraceConsumer::HandleDWMEvent(EVENT_RECORD* pEventRecord)
+    {
+        auto batchThreadEventIter = mPresentByThreadId.find(p->DriverBatchThreadId);
+        if (batchThreadEventIter != mPresentByThreadId.end() && batchThreadEventIter->second == p) {
+            mPresentByThreadId.erase(batchThreadEventIter);
+        }
+    }
+
+    // mPresentsBySubmitSequence
+    if (p->QueueSubmitSequence != 0) {
+        auto eventIter = mPresentsBySubmitSequence.find(p->QueueSubmitSequence);
+        if (eventIter != mPresentsBySubmitSequence.end() && (eventIter->second == p)) {
+            mPresentsBySubmitSequence.erase(eventIter);
+        }
+    }
+
+    // mWin32KPresentHistoryTokens
+    if (p->CompositionSurfaceLuid != 0) {
+        PMTraceConsumer::Win32KPresentHistoryTokenKey key(
+            p->CompositionSurfaceLuid,
+            p->Win32KPresentCount,
+            p->Win32KBindId
+        );
+
+        auto eventIter = mWin32KPresentHistoryTokens.find(key);
+        if (eventIter != mWin32KPresentHistoryTokens.end() && (eventIter->second == p)) {
+            mWin32KPresentHistoryTokens.erase(eventIter);
+        }
+    }
+
+    // mDxgKrnlPresentHistoryTokens
+    if (p->TokenPtr != 0) {
+        auto eventIter = mDxgKrnlPresentHistoryTokens.find(p->TokenPtr);
+        if (eventIter != mDxgKrnlPresentHistoryTokens.end() && eventIter->second == p) {
+            mDxgKrnlPresentHistoryTokens.erase(eventIter);
+        }
+    }
+
+    // mBltsByDxgContext
+    if (p->DxgKrnlHContext != 0) {
+        auto eventIter = mBltsByDxgContext.find(p->DxgKrnlHContext);
+        if (eventIter != mBltsByDxgContext.end() && eventIter->second == p) {
+            mBltsByDxgContext.erase(eventIter);
+        }
+    }
+
+    // mLastWindowPresent
+    // 0 is a invalid hwnd
+    if (p->Hwnd != 0) {
+        auto eventIter = mLastWindowPresent.find(p->Hwnd);
+        if (eventIter != mLastWindowPresent.end() && eventIter->second == p) {
+            mLastWindowPresent.erase(eventIter);
+        }
+    }
+
+    // mPresentsWaitingForDWM
+    if (p->PresentInDwmWaitingStruct) {
+        for (auto presentIter = mPresentsWaitingForDWM.begin(); presentIter != mPresentsWaitingForDWM.end(); presentIter++) {
+            // This loop should in theory be short because the present is old.
+            // If we are in this loop for dozens of times, something is likely wrong.
+            if (p == *presentIter)
+            {
+                mPresentsWaitingForDWM.erase(presentIter);
+                break;
+            }
+        }
+    }
+
+    // mPresentsByLegacyBlitToken
+    // LegacyTokenData cannot be 0 if it's in mPresentsByLegacyBlitToken list.
+    if (p->LegacyBlitTokenData != 0) {
+        auto eventIter = mPresentsByLegacyBlitToken.find(p->LegacyBlitTokenData);
+        if (eventIter != mPresentsByLegacyBlitToken.end() && eventIter->second == p) {
+            mPresentsByLegacyBlitToken.erase(eventIter);
+        }
+    }
+}
+
+void PMTraceConsumer::RemoveLostPresent(std::shared_ptr<PresentEvent> p)

```

Comment:
This looks similar enough to CompletePresent(), especially if we move to the similar modle of dequeing lost presents too.  I think we should combine these functions with a parameter for Complete vs Lost.

Reply:
I think they are similar conceptually, but if merged may turn into two big if statements or several smaller if statements on the flag condition, which might not be a huge improvement over the current model. Orienting towards not merge them.